Add Expressive Code `<Code>` component to Starlight (#1395)
withastro/starlight: docs/src/content/docs/guides/components.mdx

A list of all available icons is shown below with their associated names. Click an icon to copy the component code for it.

<IconsList />

### Code

> Starlight uses [Expressive Code](https://expressive-code.com/) to render code blocks with extended formatting possibilities. In addition to fenced code blocks in markdown & MDX documents, Expressive Code can display syntax-highlighted code snippets using its [`<Code>` component](https://expressive-code.com/key-features/code-component/).
> 
> The `<Code>` component provides props like `code`, `lang` or `meta` that allow you to dynamically define a code block's contents. Using this component, you can render code blocks from variables or data coming from external sources like files, databases or APIs.

Reviewer: Suggesting a couple changes here to link to the MD/MDX docs and switch up some of the order:

```suggestion
Starlight uses Expressive Code to render [Markdown and MDX code blocks](/guides/authoring-content/#code-blocks). Use the `<Code>` component to render syntax highlighted code when using a Markdown code block is not possible, for example, to render data coming from external sources like files, databases, or APIs.

See the [Expressive Code “Code Component” docs](https://expressive-code.com/key-features/code-component/) for full details of the props `<Code>` supports.
```

```mdx
# src/content/docs/example.mdx

import { Code } from '@astrojs/starlight/components';

export const exampleCode = `console.log('This could come from a file or CMS!');`;
export const fileName = 'example.js';
export const highlights = ["file", "CMS"];

<Code code={exampleCode} lang="js" title={fileName} mark={highlights} />
```

The code above generates the following on the page:

import { Code } from '@astrojs/starlight/components';

export const exampleCode = `console.log('This could come from a file or CMS!');`;
export const fileName = 'example.js';
export const highlights = ["file", "CMS"];

> <Code code={exampleCode} lang="js" title={fileName} mark={highlights} />

Reviewer: Since it came up several times recently, what do you think of adding something like this to explicitly document the import version of usage?

````suggestion
<Code code={exampleCode} lang="js" title={fileName} mark={highlights} />

#### Code transclusion

Use [Vite’s `?raw` import suffix](https://vitejs.dev/guide/assets#importing-asset-as-string) to import any code file as a string.
You can then pass this imported string to the `<Code>` component to include it on your page.

```mdx title="src/content/docs/example.mdx" "?raw"
import { Code } from '@astrojs/starlight/components';
import importedCode from '/src/env.d.ts?raw';

<Code code={importedCode} lang="ts" title="src/env.d.ts" />
```

The code above generates the following on the page:

import importedCode from '/src/env.d.ts?raw';

<Code code={importedCode} lang="ts" title="src/env.d.ts" />
````

Author: Love it! Great idea to add this example. Btw, this is the first time I've read the word "transclusion", I'm just trusting you that this is the correct term. :)

Reviewer: Haha, I honestly only learned it after working on Starlight and seeing it used by technical writers and some other tools. It’s a bit specialised (so I might double check with Sarah if she thinks there’s a better way to say it), but basically a technical term meaning including other content by reference in a page ([Wikipedia](https://en.wikipedia.org/wiki/Transclusion)).

A list of all available icons is shown below with their associated names. Click an icon to copy the component code for it.

<IconsList />

### Code

> Starlight uses Expressive Code to render [Markdown and MDX code blocks](/guides/authoring-content/#code-blocks). Use the `<Code>` component to render syntax highlighted code when using a Markdown code block is not possible, for example, to render data coming from external sources like files, databases, or APIs.

Reviewer 2: ```suggestion
Use the `<Code>` component to render syntax highlighted code when using a [Markdown code block](/guides/authoring-content/#code-blocks) is not possible, for example, to render data coming from external sources like files, databases, or APIs.
```
I'm having a hard time putting my finger on it, but it seemed weird to state Expressive Code here without linking to it (but then there was immediately an important link), and the fact it was at the very beginning of the sentence/introduction seemed like it was "the main point."

I think this might get to the point, and still retain all the useful information?

See the [Expressive Code “Code Component” docs](https://expressive-code.com/key-features/code-component/) for full details of the props `<Code>` supports.

```mdx
# src/content/docs/example.mdx

import { Code } from '@astrojs/starlight/components';

export const exampleCode = `console.log('This could come from a file or CMS!');`;
export const fileName = 'example.js';
export const highlights = ["file", "CMS"];

<Code code={exampleCode} lang="js" title={fileName} mark={highlights} />
```

The code above generates the following on the page:

import { Code } from '@astrojs/starlight/components';

export const exampleCode = `console.log('This could come from a file or CMS!');`;
export const fileName = 'example.js';
export const highlights = ["file", "CMS"];

<Code code={exampleCode} lang="js" title={fileName} mark={highlights} />

> #### Code transclusion

Reviewer 2: ```suggestion
#### Imported Code
```
I don't hate "transclusion"! (though, I also didn't know what it was!) ... but wonder whether something more funtional here like "Imported Code" might help, since that's in fact how it's talked about in the explanation.

If you're going to say, "you wouldn't *always* do this with code you import" then I would reply that the wording below can be shifted to talk about the process of importing code, and then the part below becomes an "and if you want it imported as a string..." description.

Reviewer: Yup I’m happy to use a simpler and more common word here too. I’d seen a few people ask us about this and use “transclusion”, but VitePress’s docs also say [“Import Code Snippets”](https://vitepress.dev/guide/markdown#import-code-snippets) for their equivalent feature if we need more convincing.
